File: Week03/LeetCode-1356.py
from functools import cmp_to_key

def noOneBits(n):
    count = 0
    while n > 0:
        if n & 0x1 == 1:
            count += 1
        n >>= 1                
    return count

def myComparator(one, two):
    count1 = noOneBits(one)
    count2 = noOneBits(two)

    if count1 == count2:
        return one-two
    else:
        return count1-count2

class Solution:    
    def sortByBits(self, arr: List[int]) -> List[int]:
        arr.sort(key=cmp_to_key(myComparator))
        return arr

File: Week03/LeetCode-1356.cpp
class Solution {
public:
    static int noOneBits(int n){
        int count = 0;
//        for (int i=0; i<sizeof(n)*8; i++){
        while (n){
            if (n & 0x1) count++;
            n >>= 1;
        } //end-for

        return count;
    } //end-noOneBits
    
    vector<int> sortByBits(vector<int>& arr) {
        auto myComparator = [](int one, int two){
            int count1 = noOneBits(one);
            int count2 = noOneBits(two);

            if (count1 == count2) return one < two;
            return count1 < count2;
        }; //end-myComparator

        sort(arr.begin(), arr.end(), myComparator);
        return arr;
    } // end-sortByBits
};
